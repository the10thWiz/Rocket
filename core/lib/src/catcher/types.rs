use either::Either;
use transient::{Any, CanRecoverFrom, CanTranscendTo, Downcast, Transience};
use crate::{http::Status, response::{self, Responder}, Request, Response};
#[doc(inline)]
pub use transient::{Static, Transient, TypeId, Inv};

/// Polyfill for trait upcasting to [`Any`]
pub trait AsAny<Tr: Transience>: Any<Tr> + Sealed {
    /// The actual upcast
    fn as_any(&self) -> &dyn Any<Tr>;
    /// convience typeid of the inner typeid
    fn trait_obj_typeid(&self) -> TypeId;
}

use sealed::Sealed;
mod sealed {
    use transient::{Any, Inv, Transient, TypeId};

    use super::AsAny;

    pub trait Sealed {}
    impl<'r, T: Any<Inv<'r>>> Sealed for T { }
    impl<'r, T: Any<Inv<'r>> + Transient> AsAny<Inv<'r>> for T {
        fn as_any(&self) -> &dyn Any<Inv<'r>> {
            self
        }
        fn trait_obj_typeid(&self) -> transient::TypeId {
            TypeId::of::<T>()
        }
    }
}

/// This is the core of typed catchers. If an error type (returned by
/// FromParam, FromRequest, FromForm, FromData, or Responder) implements
/// this trait, it can be caught by a typed catcher. (TODO) This trait
/// can be derived.
pub trait TypedError<'r>: AsAny<Inv<'r>> + Send + Sync + 'r {
    /// Generates a default response for this type (or forwards to a default catcher)
    #[allow(unused_variables)]
    fn respond_to(&self, request: &'r Request<'_>) -> Result<Response<'r>, Status> {
        Err(Status::InternalServerError)
    }

    /// A descriptive name of this error type. Defaults to the type name.
    fn name(&self) -> &'static str { std::any::type_name::<Self>() }

    /// The error that caused this error. Defaults to None.
    ///
    /// # Warning
    /// A typed catcher will not attempt to follow the source of an error
    /// more than once.
    fn source(&'r self) -> Option<&'r (dyn TypedError<'r> + 'r)> { None }

    /// Status code
    // TODO: This is currently only used for errors produced by Fairings
    fn status(&self) -> Status { Status::InternalServerError }
}

impl<'r> TypedError<'r> for std::convert::Infallible {  }

impl<'r> TypedError<'r> for () {  }

impl<'r> TypedError<'r> for std::io::Error {
    fn status(&self) -> Status {
        match self.kind() {
            std::io::ErrorKind::NotFound => Status::NotFound,
            _ => Status::InternalServerError,
        }
    }
}

impl<'r> TypedError<'r> for std::num::ParseIntError {}
impl<'r> TypedError<'r> for std::num::ParseFloatError {}
impl<'r> TypedError<'r> for std::string::FromUtf8Error {}

impl TypedError<'_> for Status {
    fn status(&self) -> Status { *self }
}

#[cfg(feature = "json")]
impl<'r> TypedError<'r> for serde_json::Error {}

#[cfg(feature = "msgpack")]
impl<'r> TypedError<'r> for rmp_serde::encode::Error {}
#[cfg(feature = "msgpack")]
impl<'r> TypedError<'r> for rmp_serde::decode::Error {}

// TODO: This is a hack to make any static type implement Transient
impl<'r, T: std::fmt::Debug + Send + Sync + 'static> TypedError<'r> for response::Debug<T> {
    fn respond_to(&self, request: &'r Request<'_>) -> Result<Response<'r>, Status> {
        format!("{:?}", self.0).respond_to(request).responder_error()
    }
}

impl<'r, L, R> TypedError<'r> for Either<L, R>
    where L: TypedError<'r> + Transient,
          L::Transience: CanTranscendTo<Inv<'r>>,
          R: TypedError<'r> + Transient,
          R::Transience: CanTranscendTo<Inv<'r>>,
{
    fn respond_to(&self, request: &'r Request<'_>) -> Result<Response<'r>, Status> {
        match self {
            Self::Left(v) => v.respond_to(request),
            Self::Right(v) => v.respond_to(request),
        }
    }

    fn name(&self) -> &'static str { std::any::type_name::<Self>() }

    fn source(&'r self) -> Option<&'r (dyn TypedError<'r> + 'r)> {
        match self {
            Self::Left(v) => Some(v),
            Self::Right(v) => Some(v),
        }
    }

    fn status(&self) -> Status {
        match self {
            Self::Left(v) => v.status(),
            Self::Right(v) => v.status(),
        }
    }
}

// TODO: This cannot be used as a bound on an untyped catcher to get any error type.
// This is mostly an implementation detail (and issue with double boxing) for
// the responder derive
#[derive(Transient)]
pub struct AnyError<'r>(pub Box<dyn TypedError<'r> + 'r>);

impl<'r> TypedError<'r> for AnyError<'r> {
    fn source(&'r self) -> Option<&'r (dyn TypedError<'r> + 'r)> {
        Some(self.0.as_ref())
    }
}

/// Validates that a type implements `TypedError`. Used by the `#[catch]` attribute to ensure
/// the `TypeError` is first in the diagnostics.
#[doc(hidden)]
pub fn type_id_of<'r, T: TypedError<'r> + Transient + 'r>() -> (TypeId, &'static str) {
    (TypeId::of::<T>(), std::any::type_name::<T>())
}

/// Downcast an error type to the underlying concrete type. Used by the `#[catch]` attribute.
#[doc(hidden)]
pub fn downcast<'r, T: TypedError<'r> + Transient + 'r>(v: Option<&'r dyn TypedError<'r>>) -> Option<&'r T>
    where T::Transience: CanRecoverFrom<Inv<'r>>
{
    // if v.is_none() {
    //     crate::trace::error!("No value to downcast from");
    // }
    let v = v?;
    // crate::trace::error!("Downcasting error from {}", v.name());
    v.as_any().downcast_ref()
}

/// Upcasts a value to `Box<dyn Error<'r>>`, falling back to a default if it doesn't implement
/// `Error`
#[doc(hidden)]
#[macro_export]
macro_rules! resolve_typed_catcher {
    ($T:expr) => ({
        #[allow(unused_imports)]
        use $crate::catcher::resolution::{Resolve, DefaultTypeErase, ResolvedTypedError};

        let inner = Resolve::new($T).cast();
        ResolvedTypedError {
            name: inner.as_ref().map(|e| e.name()),
            val: inner,
        }
    });
}

pub use resolve_typed_catcher;

pub mod resolution {
    use std::marker::PhantomData;

    use transient::{CanTranscendTo, Transient};

    use super::*;

    /// The *magic*.
    ///
    /// `Resolve<T>::item` for `T: Transient` is `<T as Transient>::item`.
    /// `Resolve<T>::item` for `T: !Transient` is `DefaultTypeErase::item`.
    ///
    /// This _must_ be used as `Resolve::<T>:item` for resolution to work. This
    /// is a fun, static dispatch hack for "specialization" that works because
    /// Rust prefers inherent methods over blanket trait impl methods.
    pub struct Resolve<'r, T: 'r>(T, PhantomData<&'r ()>);

    impl<'r, T: 'r> Resolve<'r, T> {
        pub fn new(val: T) -> Self {
            Self(val, PhantomData)
        }
    }

    /// Fallback trait "implementing" `Transient` for all types. This is what
    /// Rust will resolve `Resolve<T>::item` to when `T: !Transient`.
    pub trait DefaultTypeErase<'r>: Sized {
        const SPECIALIZED: bool = false;

        fn cast(self) -> Option<Box<dyn TypedError<'r>>> { None }
    }

    impl<'r, T: 'r> DefaultTypeErase<'r> for Resolve<'r, T> {}

    /// "Specialized" "implementation" of `Transient` for `T: Transient`. This is
    /// what Rust will resolve `Resolve<T>::item` to when `T: Transient`.
    impl<'r, T: TypedError<'r> + Transient> Resolve<'r, T>
        where T::Transience: CanTranscendTo<Inv<'r>>
    {
        pub const SPECIALIZED: bool = true;

        pub fn cast(self) -> Option<Box<dyn TypedError<'r>>> { Some(Box::new(self.0)) }
    }

    /// Wrapper type to hold the return type of `resolve_typed_catcher`.
    #[doc(hidden)]
    pub struct ResolvedTypedError<'r> {
        /// The return value from `TypedError::name()`, if Some
        pub name: Option<&'static str>,
        /// The upcast error, if it supports it
        pub val: Option<Box<dyn TypedError<'r> + 'r>>,
    }
}
