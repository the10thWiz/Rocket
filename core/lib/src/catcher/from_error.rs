use async_trait::async_trait;

use crate::http::Status;
use crate::outcome::Outcome;
use crate::request::FromRequest;
use crate::Request;

use crate::catcher::TypedError;

/// Trait used to extract types for an error catcher. You should
/// pretty much never implement this yourself. There are several
/// existing implementations, that should cover every need.
///
/// - [`Status`]: Extracts the HTTP status that this error is catching.
/// - [`&Request<'_>`]: Extracts a reference to the entire request that
///     triggered this error to begin with.
/// - [`T: FromRequest<'_>`]: Extracts type that implements `FromRequest`
/// - [`&dyn TypedError<'_>`]: Extracts the typed error, as a dynamic
///     trait object.
/// - [`Option<&dyn TypedError<'_>>`]: Same as previous, but succeeds even
///     if there is no typed error to extract.
///
/// [`Status`]: crate::http::Status
/// [`&Request<'_>`]: crate::request::Request
/// [`&dyn TypedError<'_>`]: crate::catcher::TypedError
/// [`Option<&dyn TypedError<'_>>`]: crate::catcher::TypedError
#[async_trait]
pub trait FromError<'r>: Sized {
    async fn from_error(
        status: Status,
        request: &'r Request<'r>,
        error: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status>;
}

#[async_trait]
impl<'r> FromError<'r> for Status {
    async fn from_error(
        status: Status,
        _r: &'r Request<'r>,
        _e: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status> {
        Ok(status)
    }
}

#[async_trait]
impl<'r> FromError<'r> for &'r Request<'r> {
    async fn from_error(
        _s: Status,
        req: &'r Request<'r>,
        _e: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status> {
        Ok(req)
    }
}

#[async_trait]
impl<'r, T: FromRequest<'r>> FromError<'r> for T {
    async fn from_error(
        _s: Status,
        req: &'r Request<'r>,
        _e: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status> {
        match T::from_request(req).await {
            Outcome::Success(val) => Ok(val),
            Outcome::Error(e) => {
                info!("Catcher guard error type: `{:?}`", e.name());
                Err(e.status())
            },
            Outcome::Forward(s) => {
                info!(status = %s, "Catcher guard forwarding");
                Err(s)
            },
        }
    }
}

#[async_trait]
impl<'r> FromError<'r> for &'r dyn TypedError<'r> {
    async fn from_error(
        _s: Status,
        _r: &'r Request<'r>,
        error: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status> {
        error.ok_or(Status::InternalServerError)
    }
}

#[async_trait]
impl<'r> FromError<'r> for Option<&'r dyn TypedError<'r>> {
    async fn from_error(
        _s: Status,
        _r: &'r Request<'r>,
        error: Option<&'r dyn TypedError<'r>>
    ) -> Result<Self, Status> {
        Ok(error)
    }
}
