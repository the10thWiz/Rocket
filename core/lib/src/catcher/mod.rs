//! Types and traits for error catchers and their handlers and return types.

mod catcher;
mod handler;
mod types;

pub use catcher::*;
pub use handler::*;
pub use types::*;
