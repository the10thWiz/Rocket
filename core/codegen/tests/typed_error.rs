#[macro_use] extern crate rocket;
use rocket::catcher::TypedError;
use rocket::http::Status;

fn boxed_error<'r>(_val: Box<dyn TypedError<'r> + 'r>) {}

#[derive(TypedError)]
pub enum Foo<'r> {
    First(String),
    Second(Vec<u8>),
    Third {
        #[error(source)]
        responder: std::io::Error,
    },
    #[error(status = 400)]
    Fourth {
        string: &'r str,
    },
}

#[test]
fn validate_foo() {
    let first = Foo::First("".into());
    assert_eq!(first.status(), Status::InternalServerError);
    assert!(first.source().is_none());
    boxed_error(Box::new(first));
    let second = Foo::Second(vec![]);
    assert_eq!(second.status(), Status::InternalServerError);
    assert!(second.source().is_none());
    boxed_error(Box::new(second));
    let third = Foo::Third {
        responder: std::io::Error::new(std::io::ErrorKind::NotFound, ""),
    };
    assert_eq!(third.status(), Status::InternalServerError);
    assert!(std::ptr::eq(
        third.source().unwrap(),
        if let Foo::Third { responder } = &third { responder } else { panic!() }
    ));
    boxed_error(Box::new(third));
    let fourth = Foo::Fourth { string: "" };
    assert_eq!(fourth.status(), Status::BadRequest);
    assert!(fourth.source().is_none());
    boxed_error(Box::new(fourth));
}

#[derive(TypedError)]
pub struct InfallibleError {
    #[error(source)]
    _inner: std::convert::Infallible,
}

#[derive(TypedError)]
pub struct StaticError {
    #[error(source)]
    inner: std::string::FromUtf8Error,
}

#[test]
fn validate_static() {
    let val = StaticError {
        inner: String::from_utf8(vec![0xFF]).unwrap_err(),
    };
    assert_eq!(val.status(), Status::InternalServerError);
    assert!(std::ptr::eq(
        val.source().unwrap(),
        &val.inner,
    ));
    boxed_error(Box::new(val));
}
