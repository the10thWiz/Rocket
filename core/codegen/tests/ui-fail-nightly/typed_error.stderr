error: only one lifetime is supported
 --> tests/ui-fail-nightly/typed_error.rs:8:14
  |
8 | struct Thing1<'a, 'b> {
  |              ^^^^^^^^
  |
note: error occurred while deriving `TypedError`
 --> tests/ui-fail-nightly/typed_error.rs:7:10
  |
7 | #[derive(TypedError)]
  |          ^^^^^^^^^^
  = note: this error originates in the derive macro `TypedError` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only one lifetime is supported
 --> tests/ui-fail-nightly/typed_error.rs:8:14
  |
8 | struct Thing1<'a, 'b> {
  |              ^^^^^^^^
  |
note: error occurred while deriving `Transient`
 --> tests/ui-fail-nightly/typed_error.rs:7:10
  |
7 | #[derive(TypedError)]
  |          ^^^^^^^^^^
  = note: this error originates in the derive macro `TypedError` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only one lifetime is supported
  --> tests/ui-fail-nightly/typed_error.rs:20:12
   |
20 | enum Thing3<'a, 'b> {
   |            ^^^^^^^^
   |
note: error occurred while deriving `TypedError`
  --> tests/ui-fail-nightly/typed_error.rs:19:10
   |
19 | #[derive(TypedError)]
   |          ^^^^^^^^^^
   = note: this error originates in the derive macro `TypedError` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only one lifetime is supported
  --> tests/ui-fail-nightly/typed_error.rs:20:12
   |
20 | enum Thing3<'a, 'b> {
   |            ^^^^^^^^
   |
note: error occurred while deriving `Transient`
  --> tests/ui-fail-nightly/typed_error.rs:19:10
   |
19 | #[derive(TypedError)]
   |          ^^^^^^^^^^
   = note: this error originates in the derive macro `TypedError` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/ui-fail-nightly/typed_error.rs:32:19
   |
32 | enum EmptyEnum { }
   |                   ^ consider adding a `main` function to `$DIR/tests/ui-fail-nightly/typed_error.rs`

error[E0277]: the trait bound `InnerNonError: TypedError<'r>` is not satisfied
  --> tests/ui-fail-nightly/typed_error.rs:15:5
   |
15 | /     #[error(source)]
16 | |     inner: InnerNonError,
   | |________________________^ the trait `TypedError<'r>` is not implemented for `InnerNonError`
   |
   = help: the following other types implement trait `TypedError<'r>`:
             <() as TypedError<'r>>
             <Accepted<R> as TypedError<'_>>
             <AnyError<'r> as TypedError<'r>>
             <BadRequest<R> as TypedError<'_>>
             <Conflict<R> as TypedError<'_>>
             <Created<R> as TypedError<'_>>
             <EmptyEnum as TypedError<'r>>
             <Errors<'r> as TypedError<'r>>
           and $N others
   = note: required for the cast from `&InnerNonError` to `&(dyn TypedError<'r> + 'r)`

error[E0277]: the trait bound `InnerNonError: TypedError<'r>` is not satisfied
  --> tests/ui-fail-nightly/typed_error.rs:27:7
   |
27 |     A(#[error(source)] InnerNonError),
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `TypedError<'r>` is not implemented for `InnerNonError`
   |
   = help: the following other types implement trait `TypedError<'r>`:
             <() as TypedError<'r>>
             <Accepted<R> as TypedError<'_>>
             <AnyError<'r> as TypedError<'r>>
             <BadRequest<R> as TypedError<'_>>
             <Conflict<R> as TypedError<'_>>
             <Created<R> as TypedError<'_>>
             <EmptyEnum as TypedError<'r>>
             <Errors<'r> as TypedError<'r>>
           and $N others
   = note: required for the cast from `&InnerNonError` to `&(dyn TypedError<'r> + 'r)`

error[E0004]: non-exhaustive patterns: type `&EmptyEnum` is non-empty
  --> tests/ui-fail-nightly/typed_error.rs:31:10
   |
31 | #[derive(TypedError)]
   |          ^^^^^^^^^^
   |
note: `EmptyEnum` defined here
  --> tests/ui-fail-nightly/typed_error.rs:32:6
   |
32 | enum EmptyEnum { }
   |      ^^^^^^^^^
   = note: the matched value is of type `&EmptyEnum`
   = note: references are always considered inhabited
   = note: this error originates in the derive macro `TypedError` (in Nightly builds, run with -Z macro-backtrace for more info)
